sinks/kafka/kafka.go

```diff
@@ -0,0 +1,303 @@
+func NewKafkaSpanSink(logger *logrus.Logger, brokers string, topic string, partitioner string, ackRequirement string, retries int, bufferBytes int, bufferMessages int, bufferDuration string, serializationFormat string, stats *statsd.Client) (*KafkaSpanSink, error) {
+
+	if topic == "" {
+		return nil, errors.New("Cannot start Kafka span sink with no span topic")
+	}
+
+	ll := logger.WithField("span_sink", "kafka")
+
+	serializer := serializationFormat
+	if serializer != "json" && serializer != "protobuf" {
+		ll.WithField("serializer", serializer).Warn("Unknown serializer, defaulting to protobuf")
+		serializer = "protobuf"
+	}
+
+	var finalBufferDuration time.Duration
+	if bufferDuration != "" {
+		var err error
+		finalBufferDuration, err = time.ParseDuration(bufferDuration)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	config, _ := newProducerConfig(ll, ackRequirement, partitioner, retries, bufferBytes, bufferMessages, finalBufferDuration)
+
+	ll.WithFields(logrus.Fields{
+		"brokers":         brokers,
+		"topic":           topic,
+		"partitioner":     partitioner,
+		"ack_requirement": ackRequirement,
+		"max_retries":     retries,
+		"buffer_bytes":    bufferBytes,
+		"buffer_messages": bufferMessages,
+		"buffer_duration": bufferDuration,
+	}).Info("Started Kafka span sink")
+
+	return &KafkaSpanSink{
+		logger:     ll,
+		statsd:     stats,
+		topic:      topic,
+		brokers:    brokers,
+		config:     config,
+		serializer: serializer,
+	}, nil
+}
+
+// Name returns the name of this sink.
+func (k *KafkaSpanSink) Name() string {
+	return "kafka"
+}
+
+// Start performs final adjustments on the sink.
+func (k *KafkaSpanSink) Start(cl *trace.Client) error {
+	producer, err := newConfiguredProducer(k.logger, k.brokers, k.config)
+	if err != nil {
+		return err
+	}
+	k.producer = producer
+	return nil
+}
+
+// Ingest takes the span and adds it to the ringbuffer.
+func (k *KafkaSpanSink) Ingest(span ssf.SSFSpan) error {
+
+	var enc sarama.Encoder
+	switch k.serializer {
+	case "json":
+		j, err := json.Marshal(span)
+		if err != nil {
+			k.logger.Error("Error marshalling span")
+			k.statsd.Count("kafka.span_marshal_error_total", 1, nil, 1.0)
+			return err
+		}
+		enc = sarama.StringEncoder(j)
+	case "protobuf":
+		p, err := proto.Marshal(&span)
+		if err != nil {
+			k.logger.Error("Error marshalling span")
+			k.statsd.Count("kafka.span_marshal_error_total", 1, nil, 1.0)
+			return err
+		}
+		enc = sarama.ByteEncoder(p)
+	default:
+		return fmt.Errorf("Unknown serialization format for encoding Kafka message: %s", k.serializer)
+	}
+
+	k.producer.Input() <- &sarama.ProducerMessage{
+		Topic: k.topic,
+		Value: enc,
+	}
+	atomic.AddInt64(&k.spansFlushed, 1)
+
+	return nil
+}
+
+// Flush emits metrics, since the spans have already been ingested and are
+// sending async.
+func (k *KafkaSpanSink) Flush() {
+
+	k.statsd.Count("kafka.spans_flushed_total", atomic.LoadInt64(&k.spansFlushed), nil, 1.0)
```

Comment:
I don't think this resets the flush counter. Might have to use `atomic.SwapInt64(&k.spansFlushed, 0)`.

Reply:
Also, I wonder how desirable it would be to wait for confirmation here; potentially a follow-up thing (would definitely be worth a TODO).